source/Tests/MessageContractAnalyzerFixture.cs

```diff
@@ -0,0 +1,573 @@
+        {
+            var source = WithOctopusTypes(@"
+namespace Octopus.Core.Features.ServerTasks.MessageContracts
+{
+    /// <summary>command</summary>
+    public class SimpleCommand: ICommand<SimpleCommand, SimpleResponse> 
+    {
+        [Optional]
+        public string? {|#0:GetOnlyProp|} { get; }
+        
+        [Optional]
+        public string? {|#1:ComputedProp|} => GetOnlyProp?.ToUpperInvariant();
+    }
+    /// <summary>response</summary>
+    public class SimpleResponse : IResponse { }
+}");
+
+            await Verify.VerifyAnalyzerAsync(source,
+                new DiagnosticResult(Descriptors.PropertiesOnMessageTypesMustBeMutable).WithLocation(0).WithArguments("GetOnlyProp"),
+                new DiagnosticResult(Descriptors.PropertiesOnMessageTypesMustBeMutable).WithLocation(1).WithArguments("ComputedProp"));
+        }
+
+        [Test]
+        public async Task RequiredPropertiesOnMessageTypesMustNotBeNullable()
+        {
+            var source = WithOctopusTypes(@"
+namespace Octopus.Core.Features.ServerTasks.MessageContracts
+{
+    /// <summary>request</summary>
+    public class SimpleRequest: IRequest<SimpleRequest, SimpleResponse> 
+    {
+        protected SimpleRequest() {}
+        public SimpleRequest(string s, int i, IComparable comp) { StringProperty = s; IntProperty = i; ComparableProperty = comp; }
+
+        [Required]
+        public string? {|#0:StringProperty|} { get; set; }
+
+        [Required]
+        public int? {|#1:IntProperty|} { get; set; }
+
+        [Required]
+        public IComparable? {|#2:ComparableProperty|} { get; set; }
+    }
+    /// <summary>response</summary>
+    public class SimpleResponse : IResponse { }
+}");
+
+            await Verify.VerifyAnalyzerAsync(source,
+                new DiagnosticResult(Descriptors.RequiredPropertiesOnMessageTypesMustNotBeNullable).WithLocation(0).WithArguments("StringProperty", "string"),
+                new DiagnosticResult(Descriptors.RequiredPropertiesOnMessageTypesMustNotBeNullable).WithLocation(1).WithArguments("IntProperty", "int"),
+                new DiagnosticResult(Descriptors.RequiredPropertiesOnMessageTypesMustNotBeNullable).WithLocation(2).WithArguments("ComparableProperty", "IComparable"));
+        }
+
+        [Test]
+        public async Task OptionalPropertiesOnMessageTypesMustBeNullable() // except collection types
+        {
+            var source = WithOctopusTypes(@"
+namespace Octopus.Core.Features.ServerTasks.MessageContracts
+{
+    /// <summary>request</summary>
+    public class SimpleRequest: IRequest<SimpleRequest, SimpleResponse> 
+    {
+        [Optional]
+        public string {|#0:StringProperty|} { get; set; }
+
+        [Optional]
+        public int {|#1:IntProperty|} { get; set; }
+
+        [Optional]
+        public string[] StringListProperty { get; set; } = null; // should NOT fire on this because optional nonnull collections are allowed
```

Comment:
should that ` = null` be there?

Reply:
It doesn't make a difference either way; This part will get revisited when we do default values properly